### PR TITLE
Code changes for performance improvements and other minor fixes

### PR DESCRIPTION
**Pull request by Mayuri-Kumar93**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -159,7 +159,7 @@ CyberSource supports Message Level Encryption (MLE) for Simple Order API. Messag
 Retry Pattern allows to retry sending a failed request and it will only work with `useHttpClient=true`. `allowRetry` flag enables the retry mechanism. 
   - Set the value of `allowRetry` parameter to "TRUE/FALSE". Then the system will retry the failed request as many times as configured by the merchant in the config parameter 'numberOfRetries'.
   - numberOfRetries parameter value should be set between 0 to 5. By default the value for numberOfRetries will be 5. User can set a delay in between the retry attempts.
```
Comment (mahendya1002) [inline README.md:161]: make it 3 as default

File: README.md [hunk 2]
```diff
@@ -159,7 +159,7 @@ CyberSource supports Message Level Encryption (MLE) for Simple Order API. Messag
...
-  - Config parameter for this property is 'retryInterval' in `cybs.property` file. The default value for 'retryInterval' parameter is 5 which means a delay of 5 seconds.
+  - Config parameter for this property is 'retryInterval' in `cybs.property` file. The default value for 'retryInterval' parameter is 5000 which means a delay of 5000 milliSeconds.
```
Comment (mahendya1002) [inline README.md:163]: 1000ms

File: java/src/main/java/com/cybersource/ws/client/Client.java [hunk 3]
```diff
@@ -91,15 +93,11 @@ public static Map runTransaction(
...
             setVersionInformation(request);
 
-            String merchantID = request.get(MERCHANT_ID);
-            if (merchantID == null) {
-                // if no merchantID is present in the request, get its
-                // value from the properties and add it to the request.
-                mc = new MerchantConfig(props, null);
-                merchantID = mc.getMerchantID();
-                request.put(MERCHANT_ID, merchantID);
+            boolean isMerchantConfigCacheEnabled = Boolean.parseBoolean(props.getProperty("merchantConfigCacheEnabled"));
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/Client.java:96]: boolean isMerchantConfigCacheEnabled = Boolean.parseBoolean(props.getProperty("merchantConfigCacheEnabled", "false"));

File: java/src/main/java/com/cybersource/ws/client/Client.java [hunk 4]
```diff
@@ -310,6 +310,51 @@ static String mapToString(Map src, boolean mask, int type) {
...
+        String keyAlias = props.getProperty(KEY_ALIAS);
+        if(keyAlias != null) {
+            return keyAlias;
+        }
+
+        return getMerchantId(request, props);
+    }
+
+    private static MerchantConfig getInstanceMap(Map<String, String> request, Properties props) throws ConfigException {
+        String key = getKeyForInstanceMap(request, props);
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/Client.java:348]: midOrKeyAlias

File: java/src/main/java/com/cybersource/ws/client/Client.java [hunk 5]
```diff
@@ -34,6 +34,7 @@
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/Client.java:37]: Do it for XmlClient.java

File: java/src/main/java/com/cybersource/ws/client/Client.java [hunk 6]
```diff
@@ -34,6 +34,7 @@
...
 
 /**
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/Client.java:39]: add all new properties in readme(there are 2 readme files, do in both).
add those properties with default values in cysb or test_cybs.properties file

File: java/src/main/java/com/cybersource/ws/client/MerchantConfig.java [hunk 7]
```diff
@@ -162,11 +183,47 @@ public int getLogMaximumSize() {
     public boolean getUseHttpClient() {
         return useHttpClient;
     }
-    
+
+    public boolean getUsePoolingHttpClient() {
+        return usePoolingHttpClient;
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/MerchantConfig.java:188]: useHttpClientWithConnectionPool

File: java/src/main/java/com/cybersource/ws/client/MerchantConfig.java [hunk 8]
```diff
@@ -321,13 +390,16 @@ public MerchantConfig(Properties _props, String _merchantID)
         useSignAndEncrypted = getBooleanProperty(merchantID, "useSignAndEncrypted", false);
         
         allowRetry  = getBooleanProperty(merchantID, "allowRetry", true);
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/MerchantConfig.java:392]: if (useHttpClient  && usePoolingHttpClient) {
            throw new ConfigException("both useHttpClient and usePoolingHttpClient cant be set to true, enable one of them");
        }

        if (useHttpClient || usePoolingHttpClient) {
            if(allowRetry) {
                numberOfRetries = getIntegerProperty(merchantID, "numberOfRetries", 5);
                if (numberOfRetries > 0)
                    retryInterval = getIntegerProperty(merchantID, "retryInterval", 5000);
                if (numberOfRetries < 1 || numberOfRetries > 5 || retryInterval < 0) {
                    throw new ConfigException("Invalid value of numberOfRetries and/or retryInterval");
                }
            }
        }

File: java/src/main/java/com/cybersource/ws/client/MerchantConfig.java [hunk 9]
```diff
@@ -321,13 +390,16 @@ public MerchantConfig(Properties _props, String _merchantID)
...
+        if (useHttpClient) {
+            if(allowRetry) {
+                numberOfRetries = getIntegerProperty(merchantID, "numberOfRetries", 5);
+                if (numberOfRetries > 0)
+                    retryInterval = getIntegerProperty(merchantID, "retryInterval", 5000);
+                if (numberOfRetries < 1 || numberOfRetries > 5 || retryInterval < 0) {
+                    throw new ConfigException("Invalid value of numberOfRetries and/or retryInterval");
+                }
             }
+            usePoolingHttpClient = getBooleanProperty(merchantID, "usePoolingHttpClient", false);
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/MerchantConfig.java:402]: move this line above

File: java/src/test/resources/test_cybs.properties [hunk 10]
```diff
@@ -1,63 +1,76 @@
-merchantID=< ---merchant_id ---->
-keysDirectory=<--- P12 key directory location -->
-keyAlias=<--- key alias name for P12 key --->
-keyPassword=<--- P12 key password --->
-targetAPIVersion=<-- API Version Number -->
+merchantID=cybs_ncg
```
Comment (mahendya1002) [inline java/src/test/resources/test_cybs.properties:1]: make sure you add just new keys

Comment (mahendya1002) [inline java/src/test/resources/test_cybs.properties:1]: the changes which are not required, revert it

File: java/src/main/java/com/cybersource/ws/client/Connection.java [hunk 11]
```diff
@@ -66,6 +66,9 @@ protected Connection(MerchantConfig mc, DocumentBuilder builder,
     public static Connection getInstance(
             MerchantConfig mc, DocumentBuilder builder, LoggerWrapper logger) {
         if (mc.getUseHttpClient()) {
+            if(mc.getUsePoolingHttpClient()) {
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/Connection.java:69]: public static Connection getInstance(
            MerchantConfig mc, DocumentBuilder builder, LoggerWrapper logger) {
        if (mc.getUseHttpClient()) {
            return new HttpClientConnection(mc, builder, logger);
        }else if(mc.getUsePoolingHttpClient()) {
            return new PoolingHttpClientConnection(mc, builder, logger);
        } else {
            // HttpClient is not set in properties file then JDKHttpURLConnection class instance.
            return new JDKHttpURLConnection(mc, builder, logger);
        }
    }

File: java/src/main/java/com/cybersource/ws/client/PoolingHttpClientConnection.java [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+                        URI uri = new URI(url);
+                        String hostname = uri.getHost();
+                        connectionManager = new PoolingHttpClientConnectionManager();
+                        connectionManager.setDefaultMaxPerRoute(mc.getDefaultMaxConnectionsPerRoute());
+                        connectionManager.setMaxTotal(mc.getPublishMaxConnections());
+                        final HttpHost httpHost = new HttpHost(hostname);
+                        connectionManager.setMaxPerRoute(new HttpRoute(httpHost), mc.getMaxConnectionsPerRoute());
+                        initHttpClient();
+                    } catch (URISyntaxException e) {
+                        logger.log(Logger.LT_FAULT, "invalid server url");
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/PoolingHttpClientConnection.java:66]: } catch (Exception e) {
            logger.log(Logger.LT_FAULT, "invalid server url");
            throw new ClientException(e, logger);
        }

File: java/src/main/java/com/cybersource/ws/client/PoolingHttpClientConnection.java [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    protected void initHttpClient() {
+        staleMonitorThread = new IdleConnectionMonitorThread(connectionManager, mc.getEvictThreadSleepTimeMs(), mc.getMaxKeepAliveTimeMs());
+        RequestConfig.Builder requestConfigBuilder = RequestConfig.custom()
+                .setSocketTimeout(mc.getSocketTimeoutMs())
+                .setConnectionRequestTimeout(mc.getConnectionRequestTimeoutMs())
+                .setConnectTimeout(mc.getConnectionTimeoutMs());
+
+        HttpClientBuilder httpClientBuilder = HttpClients.custom()
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/PoolingHttpClientConnection.java:80]: HttpClientBuilder httpClientBuilder = HttpClients.custom()
                .setKeepAliveStrategy(DefaultConnectionKeepAliveStrategy.INSTANCE)
                .setRetryHandler(new CustomRetryHandler())
                .setConnectionManager(connectionManager)
                .disableConnectionState();

File: java/src/main/java/com/cybersource/ws/client/PoolingHttpClientConnection.java [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+
+    @Override
+    void postDocument(Document request) throws IOException, TransformerException {
+        httpPost = new HttpPost(mc.getEffectiveServerURL());
+        StringEntity stringEntity = new StringEntity(documentToString(request), "UTF-8");
+        httpPost.setEntity(stringEntity);
+        httpPost.setHeader(Utility.ORIGIN_TIMESTAMP, String.valueOf(System.currentTimeMillis()));
+        logRequestHeaders();
+        httpContext = HttpClientContext.create();
+        httpResponse = httpClient.execute(httpPost, httpContext);
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/PoolingHttpClientConnection.java:104]: httpResponse = httpClient.execute(httpPost);

File: java/src/main/java/com/cybersource/ws/client/PoolingHttpClientConnection.java [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    @Override
+    void postDocument(Document request) throws IOException, TransformerException {
+        httpPost = new HttpPost(mc.getEffectiveServerURL());
+        StringEntity stringEntity = new StringEntity(documentToString(request), "UTF-8");
+        httpPost.setEntity(stringEntity);
+        httpPost.setHeader(Utility.ORIGIN_TIMESTAMP, String.valueOf(System.currentTimeMillis()));
+        logRequestHeaders();
+        httpContext = HttpClientContext.create();
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/PoolingHttpClientConnection.java:103]: remove  httpContext = HttpClientContext.create();

File: java/src/main/java/com/cybersource/ws/client/PoolingHttpClientConnection.java [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+public class PoolingHttpClientConnection extends Connection {
+    HttpPost httpPost = null;
+    CloseableHttpResponse httpResponse = null;
+    HttpClientContext httpContext;
+    static CloseableHttpClient httpClient = null;
+    private IdleConnectionMonitorThread staleMonitorThread;
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/PoolingHttpClientConnection.java:41]: make this static if going with ur code.

File: README.md (hunk 1 above)
Comment (Mayuri-Kumar93) [reply README.md:161]: done

File: README.md (hunk 2 above)
Comment (Mayuri-Kumar93) [reply README.md:163]: done

File: java/src/main/java/com/cybersource/ws/client/IdleConnectionMonitorThread.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Override
+    public void run() {
+        try {
+            while (!shutdown) {
+                synchronized (this) {
+                    long idleTime = getIdleTime();
+                    wait(getSleepTime());
+                    PoolingHttpClientConnectionManager poolConnMgr = (PoolingHttpClientConnectionManager) connMgr;
+                    System.out.println("before closing expired and idl conn, stats is "+poolConnMgr.getTotalStats());
```
Comment (mahendya1002) [inline java/src/main/java/com/cybersource/ws/client/IdleConnectionMonitorThread.java:30]: remove this after the testing

File: java/src/main/resources/cybs.properties [hunk 18]
```diff
@@ -15,17 +15,32 @@ sendToProduction=false
 sendToAkamai=true
 #serverURL=https://ics2wstesta.ic3.com/commerce/1.x/transactionProcessor
 
-#useHttpClient=false
+#To enable HttpClientConnection
+useHttpClient=false
+#To enable poolingHttpClientConnection
+useHttpClientWithConnectionPool=true
```
Comment (mahendya1002) [inline java/src/main/resources/cybs.properties:21]: keep it false bydefault

**Pull request merged**